wasm/src/language_service.rs

```diff
@@ -80,147 +81,142 @@ impl LanguageService {
-    pub items: Vec<CompletionItem>,
-}
-
-#[derive(Serialize, Deserialize)]
-#[allow(non_snake_case)] // These types propagate to JS which expects camelCase
-pub struct CompletionItem {
-    pub label: String,
-    pub sortText: Option<String>,
-    pub kind: String,
-    pub detail: Option<String>,
-    pub additionalTextEdits: Option<Vec<TextEdit>>,
-}
-
-#[wasm_bindgen(typescript_custom_section)]
-const ITextEdit: &'static str = r#"
-export interface ITextEdit {
-    range: { start: number; end: number; };
-    newText: string;
+        additionalTextEdits?: ITextEdit[];
+    }"#
 }
-"#;
 
-#[derive(Serialize, Deserialize)]
-#[allow(non_snake_case)] // These types propagate to JS which expects camelCase
-pub struct TextEdit {
-    pub range: Span,
-    pub newText: String,
+serializable_type! {
+    struct TextEdit {
+        pub range: Span,
+        pub newText: String,
+    },
+    r#"export interface ITextEdit {
+        range: ISpan;
+        newText: string;
+    }"#
 }
 
-#[wasm_bindgen(typescript_custom_section)]
-const IHover: &'static str = r#"
-export interface IHover {
-    contents: string;
-    span: { start: number; end: number }
+serializable_type! {
+    struct Hover {
+        pub contents: String,
+        pub span: Span,
+    },
+    r#"export interface IHover {
+        contents: string;
+        span: ISpan
+    }"#,
+    Hover,
+    IHover,
+    "IHover"
 }
-"#;
 
-#[derive(Serialize, Deserialize)]
-pub struct Hover {
-    pub contents: String,
-    pub span: Span,
+serializable_type! {
+    struct Definition {
+        pub source: String,
+        pub offset: u32,
+    },
+    r#"export interface IDefinition {
+        source: string;
+        offset: number;
+    }"#,
+    Definition,
+    IDefinition,
+    "IDefinition"
 }
 
-#[wasm_bindgen(typescript_custom_section)]
-const IDefinition: &'static str = r#"
-export interface IDefinition {
-    source: string;
-    offset: number;
+serializable_type! {
+    struct Span {
+        pub start: u32,
+        pub end: u32,
+    },
+    r#"export interface ISpan {
+        start: number;
+        end: number;
+    }"#
 }
-"#;
 
-#[derive(Serialize, Deserialize)]
-pub struct Definition {
-    pub source: String,
-    pub offset: u32,
-}
-
-#[derive(Serialize, Deserialize)]
-pub struct Span {
-    pub start: u32,
-    pub end: u32,
-}
-
-#[wasm_bindgen(typescript_custom_section)]
-const IDiagnostic: &'static str = r#"
-export interface IDiagnostic {
-    start_pos: number;
-    end_pos: number;
-    message: string;
-    severity: "error" | "warning" | "info"
-    code?: {
-        value: string;
-        target: string;
-    }
+serializable_type! {
+    pub(crate) struct VSDiagnostic {
+        pub start_pos: usize,
+        pub end_pos: usize,
+        pub message: String,
+        pub severity: String,
+        #[serde(skip_serializing_if = "Option::is_none")]
+        pub code: Option<VSDiagnosticCode>,
+    },
+    r#"export interface IDiagnostic {
```

Comment:
Interesting we lose the 'VS' from VSDiagnostic on this one when exporting to TypeScript. Is that deliberate or for a reason?

Reply:
Oh, that must have been a personal choice I made a while back - I really didn't like exposing the name as `IVSDiagnostic` to TypeScript, since it's really not a "VS" diagnostic - it has differences from both [`vscode.Diagnostic`](https://github.com/microsoft/qsharp/blob/76e91f14a9d330a29358b45692e7530a3ccae10d/vscode/src/diagnostics.ts#L4) and [`monaco.Diagnostic`](https://github.com/microsoft/qsharp/blob/76e91f14a9d330a29358b45692e7530a3ccae10d/playground/src/editor.tsx#L26). It's our own internal type.

But I left it as "VS" in Rust code since it differentiates it from `miette::Diagnostic`. And I guess the same type gets redefined as `VSDiagnostic` further up the stack (compare [here](https://github.com/microsoft/qsharp/blob/76e91f14a9d330a29358b45692e7530a3ccae10d/npm/src/vsdiagnostic.ts#L4) and [here](https://github.com/microsoft/qsharp/blob/76e91f14a9d330a29358b45692e7530a3ccae10d/wasm/src/language_service.rs#L196)) anyway. I don't know, it's a bit of a mess right now, but the current PR isn't making it _worse_, I think.